### PR TITLE
Move chat dialogs to bottom sheet

### LINKED ISSUE #9701: Place chat's overflow menus in bottom sheet component, tidy up long press menus.
# Feature Issue

### User Story

As a user I want to have a consistent and clear presentation of additional menu options available to me when I enter chats.

### Description

1. Replace the default action sheet with out bottom sheet component on pressing the ••• overflow menu in chats. 

**••• Private chats**

1_1 Chat  -_ Overflow

---

**••• Group chats**

Group Chat -_ Menu

---

**••• Public chats.** Here, the additional 'Fetch N number of days' can be implemented in a separate issue.

Public Chat -_ Overflow

---

2. Match the options displayed on long pressing on chat list items with these

**Long press on Private chats**

Home Tab -_ Long Press on Private Chat

---

**Long press on Group chats**

Home Tab -_ Long Press on Group Chat

---

**Long press on Public chats**

Home Tab -_ Long Press on Public Chat

---

### Acceptance Criteria

All items on the list were implemented, overflow menus use the bottom sheet component.

### Notes

Designs can be found inside the[ Chat Figma](https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=12%3A127), scattered around the 1-1, Group and Public chat pages. Don't hesitate to ask me with any questions `@maciej` on discord / status.

thank you! 💪

### PR DESCRIPTION
**Pull request by Ferossgp**
Also, add some UI/UX fixes to the bottom sheet component. Fixes #9701

### CODE REVIEW

**Comment by errorists** [comment]
tested on iOS, everything's in order except the couple issues I pointed out. the bottom sheet animation works great now too! 

✅ Fixed dragging https://github.com/status-im/status-react/issues/9846

issues:
• 'Fetch last 24 hours' is using the incorrect icon asset, should be `arrow_down` and not `tiny_arrow_down` [asset in Figma](https://www.figma.com/file/cb4p8AxLtTF3q1L6JYDnKN15/Index?node-id=4190%3A1757)

issues that need a follow up in separate PRs
• bottom sheets prevent screen interaction for a second or two after they're dismissed, I find this super infuriating.
• bottom sheets have hardcoded height
• sometimes the bottom sheet animation can miss its frames and start in the middle of animation when it's called out. doesn't seem to be the issue when it's animating after dismiss.

I made screen grabs from a video recording of how the first frame of animation looks like, the bottom sheet can start at wildly different positions of the animation, making it appear stuttering and laggy.

Screen Recording (ScreenFlow)

Screen Recording (ScreenFlow) 2

**Review by flexsurfer — APPROVED** [review]

**Review by errorists — APPROVED** [review]

**Comment by churik** [comment]
Let's wait until eth.rod fleet will be fixed

**Comment by yenda** [comment]
@Serhy eth.prod fleet is fixed so this could be tested

**Comment by Serhy** [comment]
@Ferossgp Builds https://i.diawi.com/auRBcY and https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-200124-175215-8571e7-pr-universal.apk on iOS13.3 and Android 8.1 has been tested respectively. 

### Issue 1

'Fech 24h history' item has no icon asset on iOS (OK with Android)

Screenshot 2020-01-27 at 15 08 53

### Issue 2
At the moment after any bottom collapsed, doesn't matter if it closed by tapping somewhere outside or closed by selecting one of it elements, - I have to tap on screen (like to bring back focus to app main screen) again and only after that further interaction with elements take effect. It happens with *any* bottom sheet. Please see Video and Reproduction below:

**Reproduction 1**:
1) Open status and create new account
2) When on Chats view -> tap `+` button (Start new chat/Join public chat bottom sheet menu appears)
3) Tap `Join public chat`
4) Tap inside input of `Join public chat`
**Actual result:** Nothing happens
5) Tap inside input of `Join public chat` again
**Actual result = Expected result** Cursor appear in the input field and keyboard appears

OR 
**Reproduction 2:**
1) Open status and create new account
2) Join any public chat
3) Being on public chat view tap top-right `...` menu button.
4) Tap outside bottom-sheet so it closed (or pull-down upon bottom-sheet to close it)
5) Tap inside message input field
**Actual result:** Nothing happens
6) Tap inside message input field again
**Actual result = Expected result** Cursor appear in the input field and keyboard appears

Video: 
https://drive.google.com/file/d/1efJ552WVpYc1g9FcBmiPSTT9OZILQp4f/view


### Issue 3:

Delete chat does not work for public chat, neither from bottom-sheet after long-tap  on public chat from Chats view nor from public chat's view called from `...` menu button. It's fine when delete Group or 1-1 chat though.

**Reproduction**:
1) Open status and create new account
2) Join any public chat
3) Being on public chat view tap top-right `...` menu button.
4) Tap 'Delete chat' button
5) Tap `Delete` from `Are you sure?` pop-up
Actual result: Nothing is deleted. Logcat when tapping `Delete` button: 
``` 

01-27 15:34:26.993 22905 22905 D ViewRootImpl@ce08b21[MainActivity]: ViewPostIme pointer 0
01-27 15:34:27.134 22905 22905 D ViewRootImpl@ce08b21[MainActivity]: ViewPostIme pointer 1
01-27 15:34:27.145 22905 22947 D ReactNativeJS: DEBUG [status-im.utils.handlers:32] - Handling re-frame event:  :group-chats.ui/remove-chat-confirmed
01-27 15:34:27.150 22905 22925 D OpenGLRenderer: eglDestroySurface = 0x927dbb70, 0x93cae000
01-27 15:34:27.150 22905 22905 D ViewRootImpl@ce08b21[MainActivity]: dispatchDetachedFromWindow
01-27 15:34:27.150 22905 22905 D InputEventReceiver: channel '5aaade0 im.status.ethereum.pr/im.status.ethereum.MainActivity (client)' ~ Disposing input event receiver.
01-27 15:34:27.150 22905 22905 D InputEventReceiver: channel '5aaade0 im.status.ethereum.pr/im.status.ethereum.MainActivity (client)' ~NativeInputEventReceiver.
01-27 15:34:27.155 22905 22960 E GoLog   : WARN [01-27|13:34:27.155] Served shhext_leaveGroupChat             reqid=198 t=268.923µs err="invalid chat ID: status"
01-27 15:34:27.164 22905 22947 W ReactNativeJS: WARN [status-im.ethereum.json-rpc:124] - :json-rpc/error shhext_leaveGroupChat :error {:code -32000, :message "invalid chat ID: status"} :params [nil "status"]
01-27 15:34:27.170 22905 22905 D InputTransport: Input channel destroyed: fd=89
01-27 15:34:27.171 22905 22905 D ViewRootImpl@920b56f[MainActivity]: MSG_WINDOW_FOCUS_CHANGED 1
01-27 15:34:27.175 22905 22905 E ViewRootImpl: sendUserActionEvent() returned.

```

Video: 
https://drive.google.com/file/d/1efJ552WVpYc1g9FcBmiPSTT9OZILQp4f/view (ending part)

**Comment by errorists** [comment]
@Serhy re: Issue 2. does the screen remain unresponsive permanently? during my testing I noticed the same behaviour but it only remained so for the first second or two after dismissing a bottom sheet.

**Comment by Serhy** [comment]
> @Serhy re: Issue 2. does the screen remain unresponsive permanently? during my testing I noticed the same behaviour but it only remained so for the first second or two after dismissing a bottom sheet.

Yea, you are right, @errorists 
I've got now you point above (`• bottom sheets prevent screen interaction for a second or two after they're dismissed, I find this super infuriating.`) 
Agree to look in separate PR, sure! 

However then it should not be merged in RC until then :)

Another issue I forgot to mention, (I just realised it's in RC as well but because in RC we don't have bottom-sheet as menu everywhere it could be reproduce with only one place):
On iOS if keyboard is opened on screen and bottom-sheet menu is called - it appears behind keyboard.



12323ddd (1)

**Comment by errorists** [comment]
Correct. @Ferossgp does your other work on bottom sheets fix Issue 2? 
As for this issue, can we dismiss any active inputs while opening the bottom sheet, or at least make it targeted for chat search and message input?

**Comment by Ferossgp** [comment]
@Serhy could you confirm that issue 2 is not present in current RC? 
For the keyboard, I think we could dismiss keyboard on bottom sheet open. And if there are inputs in the bottom sheet it will be opened again by the new input focus.

**Comment by Serhy** [comment]
> issue 2 is not present in current RC?

No, it's fine with RC. Issue 2 (`bottom sheets prevent screen interaction for a second... ` ) is not happening in RC. 
And that's great idea to dismiss keyboard when bottom sheet appears! So as you said: to dismiss keyboard on bottom sheet open. And if there are inputs in the bottom sheet it will be opened again by the new input focus.

**Comment by Ferossgp** [comment]
Thanks, @Serhy! All issues should be fixed in the latest commit.

**Comment by Serhy** [comment]
That looks good to me, @Ferossgp ! On iOS13 and Android8 and Android9 tested.
I'll fix e2e tests (some of them require updates)

Meanwhile, there is one weird issue I found.
NOTE: I faced it during usual surfing of app but below reproduction just to show how to reproduce that issue (could not reproduce with RC4) 
Also could not reproduce with bottom-sheets from any other place, only when it's requested from chat menu (well at some places).

When from chat view  `...` menu called rapidly several times, something brakes and top-left corner of app acts on taps like there is a menu items I requested are present. Only reproduced on Android.

**Reproduction:**
1) Open Status and create new account
2) Join public chat and navigate to it's view
3) Tap `...` menu button 5 times during 1 second
4) Tap `<` back button at the top-left 
5) Tap slightly below `<` back button

Video: https://drive.google.com/file/d/1cSBZW17IRB1WZpeLGZBBPZeQ3mipf-c-/view?usp=sharing (keep an eye on taps in video)

**Comment by Serhy** [comment]
One more issue (E2E actually failed with it):
**Issue 4** 
Message text is not copied when tapping `Copy` from the new bottom-sheet menu called upon long-tap in message bubble
Bot iOS and Android reproduced

**Reproduction**:
1) Open status and create new account
2) Join public chat (or any chat)
3) Send a text/emoji message
4) Long tap on the sent message bubble
5) Tap `Copy` from menu
6) Check device Clipboard whether copied message content is present 
**Actual result:** Message text is not copied

**Issue 5**
When long tap on message bubble in chat then along with  `Reply` and `Copy` actions there is `Share...` option too appears
Nothing happens I tap `Share...` in that case, - no sharing native device view appears.

File: src/status_im/ui/screens/chat/message/sheets.cljs [hunk 1]
```diff
@@ -23,4 +25,29 @@
        :title               :t/delete-message
        :icon                :main-icons/delete
        :accessibility-label :delete-transaction-button
-       :on-press            #(hide-sheet-and-dispatch [:chat.ui/delete-message chat-id message-id])}]]))
\ No newline at end of file
+       :on-press            #(hide-sheet-and-dispatch [:chat.ui/delete-message chat-id message-id])}]]))
+
+(defn message-long-press [{:keys [message-id text]}]
+  (fn []
```
Comment (Ferossgp) [inline src/status_im/ui/screens/chat/message/sheets.cljs]:
```suggestion
  (fn [{:keys [message-id text]}]
```

Fix the closure on old message values

File: src/status_im/ui/screens/chat/message/sheets.cljs [hunk 2]
```diff
@@ -23,4 +25,29 @@
        :title               :t/delete-message
        :icon                :main-icons/delete
        :accessibility-label :delete-transaction-button
-       :on-press            #(hide-sheet-and-dispatch [:chat.ui/delete-message chat-id message-id])}]]))
\ No newline at end of file
+       :on-press            #(hide-sheet-and-dispatch [:chat.ui/delete-message chat-id message-id])}]]))
+
+(defn message-long-press [{:keys [message-id text]}]
```
Comment (Ferossgp) [inline src/status_im/ui/screens/chat/message/sheets.cljs]:
```suggestion
(defn message-long-press []
```

**Comment by Serhy** [comment]
@Ferossgp I've added some accessibility-labels to elements, hope it OK.

**Comment by Ferossgp** [comment]
> When from chat view `...` menu called rapidly several times, something brakes and top-left corner of app acts on taps like there is a menu items I requested are present. Only reproduced on Android.

Seems like there is a race condition. Should be fixed in this PR https://github.com/status-im/status-react/pull/9915 I can merge them together to avoid that available to users.

**Comment by churik** [comment]
Last results for e2e: https://ethstatus.testrail.net/index.php?/runs/view/4926&group_by=cases:section_id&group_order=asc - failures are the same as in nightly.

Tested:
- that all previous issues mentioned by @Serhy and @errorists are fixed (still have https://github.com/status-im/status-react/pull/9868#issuecomment-578783681, but only if tap in 300 ms, and it is addressed separately in other PR)
- bottom-sheets from 1-1 chat, group chat, public chat (and all options)
- bottom-sheets for messages with options reply, copy, share for sender and recipient
- navigation to profile and group info screen from bottom-sheets
- bottom-sheets with opened keyboard, focus set in different fields

Question to @errorists and to @Ferossgp : we still have pop-up menu when tapping links in the chat.
photo_2020-02-19 11 51 20

Should it be addressed separately I believe?

Also, testing of group chats is blocked by https://github.com/status-im/status-react/issues/10054 and https://github.com/status-im/status-react/issues/10057
So if it is good enough for now (without testing group chats) and we agreed to postpone "Browse" options in separate PR, it is ready to go.

**Comment by Ferossgp** [comment]
@churik I've just looked at this modal and it is part of browser UI and it is reused in deep links for links which have `/browse/`. I think if it needs an update that it should be handled separately with the other places where it is used.

**Pull request merged**